components/common/modals/checkboxes.tsx

```diff
@@ -0,0 +1,106 @@
+import { Checkbox } from '@/common/forms'
+import { ConfirmationModal } from '@/common/modals/confirmation'
+import {
+  ComponentProps,
+  ReactNode,
+  useCallback,
+  useEffect,
+  useRef,
+  useState,
+} from 'react'
+
+export type CheckboxesModalProps<T> = {
+  title: string
+  items: T[]
+  itemCmp?: (a: T, b: T) => boolean
+  itemLabel?: (item: T) => string | ReactNode
+  label?: string
+  value?: T[]
+  required?: boolean
+  confirmButtonText?: string
+  onChange?: (value: T[]) => void
+  onConfirm?: (value: T[]) => void
+} & ComponentProps<'button'>
+
+export function CheckboxesModal<T>({
+  title,
+  value,
+  required,
+  onChange,
+  onConfirm,
+  items,
+  itemCmp = (a, b) => a === b,
+  itemLabel,
+  confirmButtonText,
+
+  // button props
+  children,
+  onClick,
+  ...buttonProps
+}: CheckboxesModalProps<T>) {
+  const dialogRef = useRef<HTMLDivElement>(null)
+  const [isOpen, setIsOpen] = useState(false)
+  const [internal, setInternal] = useState<T[]>(value ?? [])
+
+  const reset = useCallback(() => {
+    setInternal(value ?? [])
+  }, [value])
+
+  const updateValue = (newValue: T[]) => {
+    setInternal(newValue)
+    onChange?.(newValue)
+  }
+
+  const updateOpen = (isOpen: boolean) => {
+    reset()
+    setIsOpen(isOpen)
+  }
+
+  useEffect(() => {
+    reset()
+  }, [reset])
+
+  return (
+    <button
+      {...buttonProps}
+      onClick={(event) => {
+        onClick?.(event)
+        if (
+          !event.defaultPrevented &&
+          !dialogRef.current?.contains(event.target as Node)
+        ) {
+          updateOpen(true)
+        }
+      }}
+    >
+      {children}
+      <ConfirmationModal
```

Comment:
wonder if this component should be pulled out of the button? I've seen weird behavior due to nested components like this inside inline components. this is probably gonna be portaled out of the button anyways so probably doesn't make much of a diff.